src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -257,14 +281,22 @@ private BoundExpression RewriteStringConcatenationOneExpr(SyntaxNode syntax, Bou
         {
             if (loweredOperand.Type.SpecialType == SpecialType.System_String)
             {
-                // loweredOperand ?? ""
-                return _factory.Coalesce(loweredOperand, _factory.Literal(""));
+                // It it's a call to string.Concat, we already know the result is not null. Otherwise return loweredOperand ?? ""
```

Comment:
>It it's a call to string.Concat, we already know the result is not null [](start = 19, length = 71)

This appears to be incorrect.  You point out in the OP a circumstance in which `string.Concat` can return null.

Reply:
Right. The compiler will never emit `string.Concat(object)` on its own (it always emits `string.Concat(object) ?? ""`), but of course the user can compile the code `string.Concat(c) + null`, and that will get compiled to `string.Concat(c)`. 

I added this optimization to avoid `o + null + null` from being compiled to `string.Concat(o) ?? "" ?? ""` (which just looks like a bug to anyone inspecting the compiler output), and `o + o + ""` from being compiled to `string.Concat(o, o) ?? ""` (which is an optimization that the pre-Roslyn compiler does, but it seems it was lost in Roslyn).

I think there's a way to fix this cleanly. I'll put together a commit later, then please let me know whether you want to keep it with the fix, or remove it altogether.